File: scripts/refresh_bill_rules.py
#!/usr/bin/env python3
import argparse
import hashlib
import json
from datetime import datetime
from pathlib import Path

from urllib import request as urlrequest



def read_json(p: Path):
    return json.loads(p.read_text(encoding="utf-8"))


def sha256_text(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


def _json_request(url: str, method: str = "GET", body=None, cookie: str = ""):
    data = None
    if body is not None:
        data = json.dumps(body).encode("utf-8")
    req = urlrequest.Request(url, data=data, method=method)
    req.add_header("content-type", "application/json")
    if cookie:
        req.add_header("cookie", cookie)
    with urlrequest.urlopen(req, timeout=30) as resp:
        raw = resp.read().decode("utf-8")
        set_cookie = resp.headers.get("set-cookie", "")
        c = set_cookie.split(";")[0] if set_cookie else ""
        return (json.loads(raw) if raw else {}), c


def fetch_rules(base_url: str, username: str, password: str, tenant_id: str):
    cookie = ""
    if username and password:
        _, c = _json_request(f"{base_url}/api/auth/login", method="POST", body={"username": username, "password": password}, cookie=cookie)
        cookie = "; ".join([x for x in [cookie, c] if x])
    if tenant_id:
        _, c = _json_request(f"{base_url}/api/tenant/select", method="POST", body={"tenantId": tenant_id}, cookie=cookie)
        cookie = "; ".join([x for x in [cookie, c] if x])
    data, _ = _json_request(f"{base_url}/api/bill-rules", method="GET", cookie=cookie)
    return data


def main():
    ap = argparse.ArgumentParser(description="Manual refresh of Collector bill-rules with snapshot + manifest")
    ap.add_argument("--config", required=True, help="company config json (base_url/username/password/tenant_id)")
    ap.add_argument("--out-dir", required=True, help="snapshot directory")
    ap.add_argument("--set-latest", default="true", help="write latest.json pointer copy (true/false)")
    ap.add_argument("--dry-run", action="store_true", help="fetch+compare hash only, do not write snapshot")
    ap.add_argument("--tenant-id", help="optional tenant override")
    args = ap.parse_args()

    cfg = read_json(Path(args.config))
    base_url = str(cfg.get("base_url", "https://qb.uudc.us")).rstrip("/")
    username = cfg.get("username", "")
    password = cfg.get("password", "")
    tenant_id = args.tenant_id or cfg.get("tenant_id", "")

    out_dir = Path(args.out_dir).resolve()
    out_dir.mkdir(parents=True, exist_ok=True)
    latest_path = out_dir / "latest.json"
    manifest_path = out_dir / "manifest.json"

    rules = fetch_rules(base_url, username, password, tenant_id)
    payload = json.dumps(rules, ensure_ascii=False, sort_keys=True)
    new_hash = sha256_text(payload)

    old_hash = None
    if latest_path.exists():
        old_payload = latest_path.read_text(encoding="utf-8")
        old_hash = sha256_text(old_payload)

    changed = (new_hash != old_hash)

    if args.dry_run:
        print(json.dumps({
            "ok": True,
            "mode": "dry_run",
            "changed": changed,
            "old_hash": old_hash,
            "new_hash": new_hash,
            "tenant_id": tenant_id,
            "base_url": base_url,
        }, ensure_ascii=False, indent=2))
        return

    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    snap_path = out_dir / f"bill-rules.{stamp}.json"
    snap_path.write_text(json.dumps(rules, ensure_ascii=False, indent=2), encoding="utf-8")

    if str(args.set_latest).lower() == "true":
        latest_path.write_text(json.dumps(rules, ensure_ascii=False, indent=2), encoding="utf-8")

    manifest = {
        "updated_at": datetime.now().isoformat(timespec="seconds"),
        "tenant_id": tenant_id,
        "base_url": base_url,
        "latest_snapshot": str(snap_path),
        "latest_hash": new_hash,
        "previous_hash": old_hash,
        "changed": changed,
        "set_latest": str(args.set_latest).lower() == "true",
    }
    manifest_path.write_text(json.dumps(manifest, ensure_ascii=False, indent=2), encoding="utf-8")

    print(json.dumps({
        "ok": True,
        "changed": changed,
        "snapshot": str(snap_path),
        "latest": str(latest_path) if latest_path.exists() else None,
        "manifest": str(manifest_path),
        "hash": new_hash,
    }, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/refresh_bill_rules.mjs
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { parseArgs, readJson, createSession } from './_client.mjs';

function sha256(text) {
  return crypto.createHash('sha256').update(text, 'utf8').digest('hex');
}

function normTaxLabels(raw) {
  const q = (raw?.rules || raw)?.qboOptionDictionaries || {};
  const arr = (q.taxCodes || []).map((x) => (x?.label || x?.key || String(x || ''))).filter(Boolean).map((x) => String(x).toLowerCase());
  const s = arr.join(' | ');
  return {
    wht_out_of_scope_found: /wht\s*-\s*out\s*of\s*scope/.test(s),
    vat_12_found: /vat\s*12%/.test(s),
    non_taxable_found: /non-?taxable/.test(s),
  };
}

const args = parseArgs(process.argv);
if (!args.config || !args['out-dir']) {
  throw new Error('usage: node scripts/refresh_bill_rules.mjs --config <company.json> --out-dir <rules_cache_dir> [--set-latest true] [--dry-run true] [--tenant-id <id>]');
}

const cfg = readJson(args.config);
if (args['tenant-id']) cfg.tenant_id = args['tenant-id'];
const base = cfg.base_url || 'https://qb.uudc.us';
const outDir = path.resolve(args['out-dir']);
const setLatest = String(args['set-latest'] || 'true').toLowerCase() === 'true';
const dryRun = String(args['dry-run'] || 'false').toLowerCase() === 'true';

fs.mkdirSync(outDir, { recursive: true });
const latestPath = path.join(outDir, 'latest.json');
const manifestPath = path.join(outDir, 'manifest.json');

const api = await createSession(base, cfg);
const rules = await api.getJson('/api/bill-rules');
const payload = JSON.stringify(rules, null, 2);
const newHash = sha256(payload);
const oldHash = fs.existsSync(latestPath) ? sha256(fs.readFileSync(latestPath, 'utf8')) : null;
const changed = newHash !== oldHash;

if (dryRun) {
  console.log(JSON.stringify({ ok: true, mode: 'dry_run', changed, old_hash: oldHash, new_hash: newHash, tenant_id: cfg.tenant_id || null, base_url: base }, null, 2));
  process.exit(0);
}

const stamp = new Date().toISOString().replace(/[-:]/g, '').slice(0, 15).replace('T', '-');
const snapPath = path.join(outDir, `bill-rules.${stamp}.json`);
fs.writeFileSync(snapPath, payload, 'utf8');
if (setLatest) fs.writeFileSync(latestPath, payload, 'utf8');

const manifest = {
  updated_at: new Date().toISOString(),
  tenant_id: cfg.tenant_id || null,
  base_url: base,
  latest_snapshot: snapPath,
  latest_hash: newHash,
  previous_hash: oldHash,
  changed,
  set_latest: setLatest,
  source: 'manual_refresh',
  taxonomy_check: normTaxLabels(rules),
};
fs.writeFileSync(manifestPath, JSON.stringify(manifest, null, 2), 'utf8');

console.log(JSON.stringify({ ok: true, changed, snapshot: snapPath, latest: setLatest ? latestPath : null, manifest: manifestPath, hash: newHash, taxonomy_check: manifest.taxonomy_check }, null, 2));
